src/main/c/NativePtr.cpp

```diff
@@ -0,0 +1,352 @@
+#include "hail/NativePtr.h"
+#include "hail/NativeObj.h"
+#include <jni.h>
+#include <assert.h>
+#include <stdio.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+using hail::NativeObjPtr;
+using hail::NativeObj;
+
+NAMESPACE_BEGIN(hail)
+
+//
+// JNI interface functions corresponding to Scala NativeBase and NativePtr
+//
+
+#define NATIVEFUNC(typ) \
+  extern "C" JNIEXPORT typ JNICALL
+
+#define volatile
```

Comment:
Ah, I see.  Can this ever be safe because shared_ptr isn't POD?  Too bad shared_ptr doesn't expose the internals.  Yay, abstraction.

You can also use http://en.cppreference.com/w/cpp/types/aligned_storage.  Seems like you also want a static_assert somewhere that sizeof shared_ptr <= 2 * sizeof long.

Reply:
I think the standard-approved way would be to take sizeof(std::shared_ptr<NativeObj>), and then have the Scala hold an Array[Byte] which is precisely a copy of those bytes, and smuggle across the Scala/C++ border by doing a memcpy().  But ... we're only interested in machines with sizeof(long) = sizeof(void*) = 8 bytes; and shared_ptr has been a pair of pointers for 10 years or so ... so I'm cutting a corner by assuming those implementation details, with the fallback position that *if* we needed to run on a platform without those properties, then the porting effort would be restricted to NativePtr.scala and NativePtr.cpp.

And by also assuming the the first pointer in the pair is the object address, we get faster object access from Scala in the (usual for Hail) case of dereferencing without doing anything that might change ref-counts.